qa/L0_metrics/histogram_metrics_test.py

```diff
@@ -0,0 +1,171 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import os
+import re
+import sys
+import time
+import unittest
+from functools import partial
+
+import numpy as np
+import requests
+import tritonclient.grpc as grpcclient
+from tritonclient.utils import InferenceServerException
+
+sys.path.append("../common")
+import test_util as tu
+
+MILLIS_PER_SEC = 1000
+
+
+def get_histogram_metric_key(
+    metric_family, model_name, model_version, metric_type, le=""
+):
+    if metric_type in ["count", "sum"]:
+        return f'{metric_family}_{metric_type}{{model="{model_name}",version="{model_version}"}}'
+    elif metric_type == "bucket":
+        return f'{metric_family}_{metric_type}{{model="{model_name}",version="{model_version}",le="{le}"}}'
+    else:
+        return None
+
+
+class TestHistogramMetrics(tu.TestResultCollector):
+    def setUp(self):
+        self.tritonserver_ipaddr = os.environ.get("TRITONSERVER_IPADDR", "localhost")
+
+    def get_histogram_metrics(self, metric_family: str):
+        r = requests.get(f"http://{self.tritonserver_ipaddr}:8002/metrics")
+        r.raise_for_status()
+
+        # Regular expression to match the pattern
+        pattern = f"^{metric_family}.*"
+        histogram_dict = {}
+
+        # Find all matches in the text
+        matches = re.findall(pattern, r.text, re.MULTILINE)
+
+        for match in matches:
+            key, value = match.rsplit(" ")
+            histogram_dict[key] = int(value)
+
+        return histogram_dict
+
+    def async_stream_infer(self, model_name, inputs, outputs):
+        try:
+            triton_client = grpcclient.InferenceServerClient(url="localhost:8001")
+        except Exception as e:
+            print("context creation failed: " + str(e))
+            sys.exit()
+
+        # Define the callback function. Note the last two parameters should be
+        # result and error. InferenceServerClient would povide the results of an
+        # inference as grpcclient.InferResult in result. For successful
+        # inference, error will be None, otherwise it will be an object of
+        # tritonclientutils.InferenceServerException holding the error details
+        def callback(user_data, result, error):
+            if error:
+                user_data.append(error)
+            else:
+                user_data.append(result)
+
+        # list to hold the results of inference.
+        user_data = []
+
+        # Inference call
+        triton_client.start_stream(callback=partial(callback, user_data))
+        triton_client.async_stream_infer(
+            model_name=model_name,
+            inputs=inputs,
+            outputs=outputs,
+        )
+        triton_client.stop_stream()
+
+        # Wait until the results are available in user_data
+        time_out = 10
+        while (len(user_data) == 0) and time_out > 0:
+            time_out = time_out - 1
+            time.sleep(1)
```

Comment:
Are we only expecting one response from the request?

If I understand correctly, the ensemble is to simulate a LLM with a 
postprocessing step. The LLM is expected to return more than one word under most if not all circumstances. Can you update the decoupled model to have it returns multiple responses? And verify the `count` and `sum` are correctly aggregated with respect to the responses per request?

Reply:
This tests aim to verify not only the ensemble model but also each decoupled sub-model is updated per request. Only the time to first response is observed to histogram so I picked one response per request.

Reply:
Also the models in this ensemble model have input and output both with shape [1]. The output is a single value which feeds to the next model input as the wait seconds in `execute` function, which is used to test `sum` value.

Reply:
> Only the time to first response is observed to histogram so I picked one response per request.

Can you verify the time to first response is not updated if a second response is returned?

Reply:
It will take a while to update the models. Expect update on Monday if there's no blocker.